FWCore/TestModules/plugins/CheckClassInfo.cc

```diff
@@ -0,0 +1,153 @@
+
+  class CheckClassInfo : public edm::global::EDAnalyzer<> {
+  public:
+    explicit CheckClassInfo(edm::ParameterSet const&);
+    ~CheckClassInfo() override = default;
+
+    void analyze(edm::StreamID, edm::Event const&, edm::EventSetup const&) const override {}
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  private:
+    void check(edm::ProductDescription const& product) const;
+
+    std::vector<edm::ProductNamePattern> eventProducts_;
+    std::vector<edm::ProductNamePattern> lumiProducts_;
+    std::vector<edm::ProductNamePattern> runProducts_;
+    std::vector<edm::ProductNamePattern> processProducts_;
+  };
+
+  CheckClassInfo::CheckClassInfo(edm::ParameterSet const& config)
+      : eventProducts_(edm::productPatterns(config.getUntrackedParameter<std::vector<std::string>>("eventProducts"))),
+        lumiProducts_(edm::productPatterns(config.getUntrackedParameter<std::vector<std::string>>("lumiProducts"))),
+        runProducts_(edm::productPatterns(config.getUntrackedParameter<std::vector<std::string>>("runProducts"))),
+        processProducts_(
+            edm::productPatterns(config.getUntrackedParameter<std::vector<std::string>>("processProducts"))) {
+    callWhenNewProductsRegistered([this](edm::ProductDescription const& product) {
+      switch (product.branchType()) {
+        case edm::InEvent:
+          for (auto const& label : eventProducts_)
+            if (label.match(product)) {
+              check(product);
+              break;
+            }
+          break;
+
+        case edm::InLumi:
+          for (auto const& label : lumiProducts_)
+            if (label.match(product)) {
+              check(product);
+              break;
+            }
+          break;
+
+        case edm::InRun:
+          for (auto const& label : runProducts_)
+            if (label.match(product)) {
+              check(product);
+              break;
+            }
+          break;
+
+        case edm::InProcess:
+          for (auto const& label : processProducts_)
+            if (label.match(product)) {
+              check(product);
+              break;
+            }
+          break;
+
+        default:
+          throw edm::Exception(edm::errors::LogicError)
+              << "Unexpected branch type " << product.branchType() << "\nPlease contact a Framework developer\n";
+      }
+    });
+  }
+
+  void CheckClassInfo::check(edm::ProductDescription const& product) const {
+    if (product.transient()) {
+      edm::LogVerbatim("CheckClassInfo") << "The product " << product.friendlyClassName() << '_'
+                                         << product.moduleLabel() << '_' << product.productInstanceName() << '_'
+                                         << product.processName() << " is transient, and will not be queried.\n";
+    } else {
+      edm::LogVerbatim log("CheckClassInfo");
+      std::cerr << "product " << product.friendlyClassName() << '_' << product.moduleLabel() << '_'
```

Comment:
The printouts should use MessageLogger instead of `std::cerr`. I noticed though the `log` object is currently unused, did you perhaps intend to use it instead of `std::cerr`?

Reply:
yes, though I have to check that the MessageLogger message does not get lost in case of problems.

Reply:
> yes, though I have to check that the MessageLogger message does not get lost in case of problems.

Reporting the problem cases via exception (with message containing the relevant information) would make the problem reporting independent of MessageLogger's behavior.

Reply:
ok